lib/srv/regular/sshserver_test.go

```diff
@@ -1146,6 +1147,222 @@ func (s *SrvSuite) TestGlobalRequestRecordingProxy(c *C) {
+	netConn, err := r.listener.Accept()
+	if err != nil {
+		return nil, nil, nil, trace.Wrap(err)
+	}
+	srvConn, chs, reqs, err := ssh.NewServerConn(netConn, &r.cfg)
+	if err != nil {
+		netConn.Close()
+		return nil, nil, nil, trace.Wrap(err)
+	}
+	return srvConn, chs, reqs, nil
+}
+
+func (r *rawNode) Close() error {
+	return trace.Wrap(r.listener.Close())
+}
+
+// newRawNode constructs a new raw node instance.
+func (s *SrvSuite) newRawNode(c *C) *rawNode {
+	hostname, err := os.Hostname()
+	c.Assert(err, IsNil)
+
+	// Create host key and certificate for node.
+	keys, err := s.server.Auth().GenerateServerKeys(auth.GenerateServerKeysRequest{
+		HostID:               "raw-node",
+		NodeName:             "raw-node",
+		Roles:                teleport.Roles{teleport.RoleNode},
+		AdditionalPrincipals: []string{hostname},
+		DNSNames:             []string{hostname},
+	})
+	c.Assert(err, IsNil)
+
+	signer, err := sshutils.NewSigner(keys.Key, keys.Cert)
+	c.Assert(err, IsNil)
+
+	// configure a server which allows any client to connect
+	cfg := ssh.ServerConfig{
+		NoClientAuth: true,
+		PasswordCallback: func(conn ssh.ConnMetadata, password []byte) (*ssh.Permissions, error) {
+			return &ssh.Permissions{}, nil
+		},
+		PublicKeyCallback: func(conn ssh.ConnMetadata, key ssh.PublicKey) (*ssh.Permissions, error) {
+			return &ssh.Permissions{}, nil
+		},
+	}
+	cfg.AddHostKey(signer)
+
+	listener, err := net.Listen("tcp", ":0")
+	c.Assert(err, IsNil)
+
+	_, port, err := net.SplitHostPort(listener.Addr().String())
+	c.Assert(err, IsNil)
+
+	addr := net.JoinHostPort(hostname, port)
+
+	return &rawNode{
+		listener: listener,
+		cfg:      cfg,
+		addr:     addr,
+	}
+}
+
+// startX11EchoServer starts a fake node which, for each incomging SSH connection, accepts an
+// X11 forwarding request and then dials a single X11 channel which echoes all bytes written
+// to it.  Used to verify the behavior of X11 forwarding in recording proxies.
+func (s *SrvSuite) startX11EchoServer(ctx context.Context, c *C) *rawNode {
+	node := s.newRawNode(c)
+	go func() {
+		ctx, cancel := context.WithCancel(ctx)
+		defer cancel()
+		for {
+			conn, chs, _, err := node.accept()
+			if err != nil {
+				log.Warnf("X11 echo server closing: %v", err)
+				return
+			}
+			go func() {
+				defer conn.Close()
+
+				// expect client to open a session channel
+				var nch ssh.NewChannel
+				select {
+				case nch = <-chs:
+				case <-time.After(time.Second * 3):
+					c.Fatalf("Timeout waiting for session channel")
+				case <-ctx.Done():
+					return
+				}
+				c.Assert(nch.ChannelType(), Equals, teleport.ChanSession)
+
+				sch, creqs, err := nch.Accept()
+				c.Assert(err, IsNil)
+				defer sch.Close()
+
+				// expect client to send an X11 forwarding request
+				var req *ssh.Request
+				select {
+				case req = <-creqs:
+				case <-time.After(time.Second * 3):
+					c.Fatalf("Timeout waiting for x11 forwarding request")
```

Comment:
`X11`?

Reply:
This is deliberate.  I decided to follow the capitalization scheme that openssh uses for `X11`/`x11` (and what teleport uses for things like `HTTP`)... where it is lowercase in logs, errors, configs, etc.